README.pod
=encoding utf8


This is Perl module B<App::FromUnixtime>.

=begin html

<a href="http://travis-ci.org/bayashi/App-FromUnixtime"><img src="https://secure.travis-ci.org/bayashi/App-FromUnixtime.png"/></a> <a href="https://coveralls.io/r/bayashi/App-FromUnixtime"><img src="https://coveralls.io/repos/bayashi/App-FromUnixtime/badge.png?branch=master"/></a>

=end html


=head1 INSTALLATION

App::FromUnixtime installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan App::FromUnixtime

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install


=head1 DOCUMENTATION

App::FromUnixtime documentation is available as in POD. So you can do:

    % perldoc App::FromUnixtime


=head1 REPOSITORY

App::FromUnixtime is hosted on github
L<http://github.com/bayashi/App-FromUnixtime>


=head1 LICENSE

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 AUTHOR

Dai Okabayashi E<lt>bayashi@cpan.orgE<gt>


Changes
Revision history for Perl extension App::FromUnixtime

0.10    Sun Jan  4 20:09:09 JST 2015
        - also export 'from_unixtime' function for one liner (thanks to tokuhirom)
        - add '--no-re' option
        - rename internal function: _from_unixtime -> _replace_unixtime
        - fix pod

0.09    Sat Jan  3 21:38:27 JST 2015
        - add tests

0.08    Sat Jan  3 19:49:17 JST 2015
        - refactor
        - enable to replace unixtime args like `from_unixtime 1419702037`
        - enhance POD

0.07    Fri Jan  2 17:07:35 JST 2015
        - only replace after keyword
        - enable to replace only unixtime
        - remove eval

0.06    Mon Dec 29 21:21:59 JST 2014
        - enhance POD

0.05    Mon Dec 29 16:17:03 JST 2014
        - add pod for --re option

0.04    Mon Dec 29 15:49:52 JST 2014
        - add re option
        - refactor for metrics

0.03    Sun Dec 28 21:16:57 JST 2014
        - fix dependency

0.02    Sun Dec 28 19:16:11 JST 2014
        - use custom configuration file: .from_unixtimerc
        - add new options: --start-bracket, --end-bracket
        - add regexp pattarn for column: '_time'
        - enhanced tests

0.01    Sun Dec 28 12:13:57 JST 2014
        - original version


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    name                => 'App-FromUnixtime',
    module_name         => 'App::FromUnixtime',
    license             => 'perl',
    dist_author         => 'Dai Okabayashi <bayashi@cpan.org>',
    dist_version_from   => 'lib/App/FromUnixtime.pm',
    configure_requires  => {
        'Module::Build' => 0.38,
    },
    build_requires      => {
        'Test::More' => 0.88,
        'Capture::Tiny' => 0,
    },
    requires => {
        'perl' => '5.008001',
        'strict' => 0,
        'warnings' => 0,
        'Getopt::Long' => 0,
        'IO::Interactive::Tiny' => 0,
        'Pod::Usage' => 0,
        'POSIX' => 0,
        'Config::CmdRC' => 0,
        'Exporter' => 0,
    },
    script_files        => [ 'script/from_unixtime' ],
    add_to_cleanup      => [ 'App-FromUnixtime-*' ],
    meta_merge     => {
      resources => {
        repository  => 'http://github.com/bayashi/App-FromUnixtime',
        bugtracker  => 'http://github.com/bayashi/App-FromUnixtime/issues',
      }
    },
    recursive_test_files => 1,
    test_files => (-d '.git' || $ENV{RELEASE_TESTING}) ? ['t/', 'xt/'] : ['t/'],
);

$builder->create_build_script();


lib/App/FromUnixtime.pm
package App::FromUnixtime;
use strict;
use warnings;
use Getopt::Long qw/GetOptionsFromArray/;
use IO::Interactive::Tiny;
use POSIX qw/strftime/;
use Config::CmdRC qw/.from_unixtimerc/;
use Exporter 'import';
our @EXPORT = qw/from_unixtime/;

our $VERSION = '0.10';

our $MAYBE_UNIXTIME = join '|', (
    'created_(?:at|on)',
    'updated_(?:at|on)',
    'date',
    'unixtime',
    '_time',
);

our $DEFAULT_DATE_FORMAT = '%a, %d %b %Y %H:%M:%S %z';

sub run {
    my $self = shift;
    my @argv = @_;

    my $config = +{};
    _get_options($config, \@argv);

    _main($config);
}

sub _main {
    my $config = shift;

    if ( ! IO::Interactive::Tiny::is_interactive(*STDIN) ) {
        while ( my $line = <STDIN> ) {
            chomp $line;
            if ( my $match = _may_replace($line, $config) ) {
                if ( ! _may_not_replace($line, $config) ) {
                    _replace_unixtime($match => \$line, $config);
                }
            }
            print "$line\n";
        }
    }
    else {
        for my $unixtime (@{$config->{unixtime}}) {
            _replace_unixtime($unixtime => \$unixtime, $config);
            print "$unixtime\n";
        }
    }
}

sub _may_replace {
    my ($line, $config) = @_;

    if ($line =~ m!(?:$MAYBE_UNIXTIME)[^\d]*(\d+)!
                || ($config->{_re} && $line =~ m!(?:$config->{_re})[^\d]*(\d+)!)
                || $line =~ m!^[\s\t\r\n]*(\d+)[\s\t\r\n]*$!
    ) {
        return $1;
    }
}

sub _may_not_replace {
    my ($line, $config) = @_;

    return unless $config->{'no-re'};

    for my $no_re (@{$config->{'no-re'}}) {
        return 1 if $line =~ m!$no_re!;
    }
}

sub _replace_unixtime {
    my ($maybe_unixtime, $line_ref, $config) = @_;

    if ($maybe_unixtime > 2**31-1) {
        return;
    }

    my $date = strftime($config->{format}, localtime($maybe_unixtime));
    my $replaced_unixtime = sprintf(
        "%s%s%s%s",
        $maybe_unixtime,
        $config->{'start-bracket'},
        $date,
        $config->{'end-bracket'},
    );

    $$line_ref =~ s/$maybe_unixtime/$replaced_unixtime/;
}

sub from_unixtime {
    my ($lines, @argv) = @_;

    my $config = +{};
    _get_options($config, \@argv);

    my @replaced_lines;
    for my $line ( split /\n/, $lines ) {
        if ( my $match = _may_replace($line, $config) ) {
            _replace_unixtime($match => \$line, $config);
        }
        push @replaced_lines, $line;
    }
    return join("\n", @replaced_lines);
}

sub _get_options {
    my ($config, $argv) = @_;

    GetOptionsFromArray(
        $argv,
        'f|format=s'      => \$config->{format},
        'start-bracket=s' => \$config->{'start-bracket'},
        'end-bracket=s'   => \$config->{'end-bracket'},
        're=s@'           => \$config->{re},
        'no-re=s@'        => \$config->{'no-re'},
        'h|help' => sub {
            _show_usage(1);
        },
        'v|version' => sub {
            print "$0 $VERSION\n";
            exit 1;
        },
    ) or _show_usage(2);

    _validate_options($config, $argv);
}

sub _validate_options {
    my ($config, $argv) = @_;

    $config->{format} ||= RC->{format} || $DEFAULT_DATE_FORMAT;
    $config->{'start-bracket'} ||= RC->{'start-bracket'} || '(';
    $config->{'end-bracket'}   ||= RC->{'end-bracket'}   || ')';
    if (ref RC->{re} eq 'ARRAY') {
        push @{$config->{re}}, @{RC->{re}};
    }
    elsif (RC->{re}) {
        push @{$config->{re}}, RC->{re};
    }
    if ($config->{re}) {
        $config->{_re} = join '|', map { quotemeta $_;  } @{$config->{re}};
    }
    push @{$config->{unixtime}}, @{$argv};
}

sub _show_usage {
    my $exitval = shift;

    require Pod::Usage;
    Pod::Usage::pod2usage(-exitval => $exitval);
}

1;

__END__

=head1 NAME

App::FromUnixtime - to convert from unixtime to date suitably


=head1 SYNOPSIS

    use App::FromUnixtime;

    App::FromUnixtime->run(@ARGV);


=head1 DESCRIPTION

C<App::FromUnixtime> provides the L<from_unixtime> command and the B<from_unixtime> function.


=head1 METHOD

=head2 run

run to convert process


=head1 EXPORT FUNCTION

=head2 from_unixtime($line, @options)

C<App::FromUnixtime> exports B<from_unixtime> function for converting the string that may be included unixtime.

    use App::FromUnixtime;

    print from_unixtime('created_at 1419702037'); # created_at 1419702037(Sun, 28 Dec 2014 02:40:37 +0900)

    print from_unixtime('created_at 1419702037', '--format' => '%Y-%m-%d'); # created_at 1419702037(2014-12-28)

See L<from_unixtime> command for more options.


=head1 REPOSITORY

App::FromUnixtime is hosted on github: L<http://github.com/bayashi/App-FromUnixtime>

I appreciate any feedback :D


=head1 AUTHOR

Dai Okabayashi E<lt>bayashi@cpan.orgE<gt>


=head1 SEE ALSO

L<from_unixtime>


=head1 LICENSE

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=cut
